cmd/dockerd/daemon.go

```diff
@@ -484,7 +484,7 @@ func initRouter(s *apiserver.Server, d *daemon.Daemon, c *cluster.Cluster) {
 		image.NewRouter(d, decoder),
 		systemrouter.NewRouter(d, c),
 		volume.NewRouter(d),
-		build.NewRouter(dockerfile.NewBuildManager(d)),
+		build.NewRouter(manager.NewBuildManager(d, d), d),
```

Comment:
If `tagSquashBackend` were passed to `NewRouter()` then the code to do the tag and squash would have to be in `NewRouter()`, right? I don't think that's the right place for it either.

I think we need to establish a few layers of what we call the "builder".

"tag" and "squash" definitely doesn't belong in the "core builder", but I don't think it should be in the router either. The router should only be reponsible for handling http concerns (parse args, and call some backend).

We need some package that isn't "builder" and isn't the router. 

With this change I was thinking that `manager` would be a higher level abstraction over what is currently "core builder". It would give us a place to build up a few diferent dependencies that we can pass along as backends to "core builder". One of these things is the tag and squash.

> HasExperimental is a daemon concept that builder(and probably even router) should not know about

I agree about the builder, but I think experimental is already a core part of routers because they change their behaviour based on this setting. There are "experimental routes". So I have moved it into the router in this PR.

>  we should open up an issue to discuss if we want to continue with squash or deprecate it 

I agree, I would be +1 to deprecate, but for now I think it helps to just extarct it from the underlying builder. If we decide to remove it later, this change just makes it easier.